OsmAnd/src/net/osmand/plus/views/layers/PointLocationLayer.java

```diff
@@ -69,6 +82,206 @@
 	private boolean nm;
 	private boolean locationOutdated;
 
+	// location pin marker
+	private static final int MARKER_ID_MY_LOCATION = 1;
+	private static final int MARKER_ID_NAVIGATION = 2;
+	private MapMarkersCollection markersCollection;
+	private MapMarker myLocationMarker;
+	private MapMarker navigationMarker;
+	private SWIGTYPE_p_void onSurfaceIconKey = null;
+	private SWIGTYPE_p_void onSurfaceHeadingIconKey = null;
+	private boolean markersNeedInvalidate = true;
+	private boolean hasHeading = false;
+	private boolean hasHeadingCached = true;
+	private Location lastKnownLocation;
+	private float lastHeading = 0.0f;
+
+	private enum MarkerState {
+		Stay,
+		Move,
+		None,
+	}
+
+	private MarkerState currentMarkerState = MarkerState.Stay;
+
+	private void setMarkerState(MarkerState markerState) {
+		if (currentMarkerState == markerState) {
+			return;
+		}
+		currentMarkerState = markerState;
+		updateMarkerState();
+	}
+
+	@Override
+	public void updateLocation(Location location) {
+		lastKnownLocation = location;
+		if (view != null && view.hasMapRenderer()) {
+			view.getApplication().runInUIThread(() -> updateMarkerData(lastKnownLocation, lastKnownLocation.getAccuracy(), null), 0);
+		}
+	}
+
+	@Override
+	public void updateCompassValue(float value) {
+		if (Math.abs(MapUtils.degreesDiff(value, lastHeading)) > MapViewTrackingUtilities.COMPASS_HEADING_THRESHOLD) {
+			lastHeading = value;
+			if (view != null && view.hasMapRenderer()) {
+				view.getApplication().runInUIThread(() -> updateMarkerData(null, null, lastHeading), 0);
+			}
+		}
+	}
+
+
+	private MapMarker recreateMarker(MapMarker oldMarker, LayerDrawable icon, int id, int baseColor) {
+		if (view == null || icon == null) {
+			return null;
+		}
+
+		if (markersCollection == null) {
+			markersCollection = new MapMarkersCollection();
+		} else if (oldMarker != null) {
+			markersCollection.removeMarker(oldMarker);
+			oldMarker.delete();
+		}
+
+		MapMarkerBuilder myLocMarkerBuilder = new MapMarkerBuilder();
+		myLocMarkerBuilder.setMarkerId(id);
+		myLocMarkerBuilder.setIsAccuracyCircleSupported(true);
+		myLocMarkerBuilder.setAccuracyCircleBaseColor(NativeUtilities.createFColorRGB(baseColor));
+		myLocMarkerBuilder.setPinIconVerticalAlignment(MapMarker.PinIconVerticalAlignment.CenterVertical);
+		myLocMarkerBuilder.setPinIconHorisontalAlignment(MapMarker.PinIconHorisontalAlignment.CenterHorizontal);
+		myLocMarkerBuilder.setIsHidden(true);
+
+		float scale = getTextScale();
+		Bitmap markerBitmap = AndroidUtils.createScaledBitmap(icon, scale);
+		if (markerBitmap != null) {
+			onSurfaceIconKey = SwigUtilities.getOnSurfaceIconKey(1);
+			myLocMarkerBuilder.addOnMapSurfaceIcon(onSurfaceIconKey,
+					NativeUtilities.createSkImageFromBitmap(markerBitmap));
+		}
+
+		if (!locationOutdated && hasHeading) {
```

Comment:
It will be better if appearing / disappearing of heading will not cause recreating markers collection (thus will not flicker)

Reply:
I absolutely agree, but as we discussed before.
Unfortunately there are no core-API to control visibility of heading Icon as a part of marker.
Thus there are two options:
1. recreate it marker with/without heading until API will be extended by missed functionality (to control visibility of single icon as a part of marker);
2. create x2 amount of marker instances of markers with/without heading; - it looks like a trick, thus I decided to use 1st option.

I'll use 2nd option in case of it suitable - just let me know.